src/Data/Conduit/Tar/Unix.hs

```diff
@@ -67,7 +67,7 @@ restoreFile FileInfo {..} = do
     case fileType of
         FTDirectory -> do
             liftIO $ do
-                Dir.createDirectoryIfMissing False fpStr
+                Dir.createDirectoryIfMissing True fpStr
```

Comment:
This introduces an unnecessary overhead. tar will always include the full tree, namely all individual directories, so we don't want to create the parents on each call. As far as the directory the archive is being extracted to, than it is responsibility of the user to make sure that directory exists.

Reply:
`createDirectoryIfMissing True` isn't any slower in the happy case where only one or zero directories need creating. It also means that if a tar file is slightly malformed instead of getting a confusing error, it happily works. I personally think that having `createDirectoryIfMissing False` is an antipattern.

Reply:
Fair enough, but even if performance has no difference, I do think that it could be beneficial to be able to detect even slightly malformed tarballs. On the other hand, I can see how it creates a usability issue, since there are a lot of problematic tarballs out there. Precisely for that reason I added lenient untarring and I'd like to suggest we use that concept in recreating the tree as well, so on this line: https://github.com/snoyberg/tar-conduit/blob/master/src/Data/Conduit/Tar/Unix.hs#L74 we do:
````
       Dir.createDirectoryIfMissing lenient fpStr
```

And the same thing when restoring hard links. @ndmitchell what are you thoughts on this?
FWIW, I also think that `createDirectoryIfMissing False` is an antipattern most of the time, but in case of tar format it kinda make sense to me.

Reply:
@lehins makes a lot of sense with the `lenient` flag. I agree that I can see the sense in `createIfMissing False` for tar format files, as an unusual case.

Reply:
I changed it accordingly.

Reply:
@gaborh-da, this assumption is the incorrect one:
> files are not restored in (the right) order.

Tar is a linear format and files will be always restored in the same order they've been added to the archive ( it was created for tapes afterall :) )
What happens in your example is that tar creates an archive with:
```
tmp/abc [Dir]
tmp/abc/a [File]
tmp/abc/b [HardLink]
```
without the top level `./tmp`, so whenever you restore the archive into `/tmp/ttt/`, the directory `/tmp/ttt/tmp` doesn't exist, thus it errors out.

Here is what I recommend we do to alleviate the problem in this case. It isn't really specific to hard links or even regular files, so let's do: `createDirectoryIfMissing True fpStr` on the `FTDirectory`, since above case will be a common one and isn't really considered a malformed tar, but the rest of the tar types will have `createDirectoryIfMissing lenient fpStr`. 

Thanks for finding this edge case ;)

Reply:
@gaborh-da On a second thought, it would be possible to create an archive with just one file and no folders: `tar cvf blooo.tar /tmp/abc/a`, which would contain only one file `tmp/abc/a`, and cause an error when restoring.

@gaborh-da and @ndmitchell, sorry for this whole `createDirectoryIfMissing False` nonsense, I guess it doesn't make sense in the tar context either, so we should use `createDirectoryIfMissing True` everywhere

@gaborh-da, if you don't feel like making those last changes, it's ok, I can do them. Let if you are done with it and I'll merge this PR. Thank you for all your work ;)

Reply:
I will use `createDirectoryIfMissing True` everywhere. :)

Reply:
Even in Windows.hs.